src/librustc_mir/transform/const_prop.rs

```diff
@@ -659,9 +661,70 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             return None;
         }
 
+        self.eval_rvalue_with_identities(rvalue, place)
+    }
+
+    // Attempt to use albegraic identities to eliminate constant expressions
+    fn eval_rvalue_with_identities(
+        &mut self,
+        rvalue: &Rvalue<'tcx>,
+        place: Place<'tcx>,
+    ) -> Option<()> {
         self.use_ecx(|this| {
-            trace!("calling eval_rvalue_into_place(rvalue = {:?}, place = {:?})", rvalue, place);
-            this.ecx.eval_rvalue_into_place(rvalue, place)?;
+            match rvalue {
+                Rvalue::BinaryOp(op, left, right) | Rvalue::CheckedBinaryOp(op, left, right) => {
+                    let l = this.ecx.eval_operand(left, None);
+                    let r = this.ecx.eval_operand(right, None);
+
+                    if let (Ok(_), Ok(_)) = (&l, &r) {
+                        return Ok(this.ecx.eval_rvalue_into_place(rvalue, place)?);
+                    }
+
+                    let const_arg = this.ecx.read_immediate(l.or(r)?)?;
+                    let dest = this.ecx.eval_place(place)?;
+
+                    match op {
+                        BinOp::BitAnd => {
+                            if const_arg.to_scalar()?.to_bool()? == false {
+                                this.ecx.write_immediate(*const_arg, dest)?;
+                            }
+                        }
+                        BinOp::BitOr => {
+                            if const_arg.to_scalar()?.to_bool()? == true {
+                                this.ecx.write_immediate(*const_arg, dest)?;
+                            }
+                        }
+                        BinOp::Mul => {
+                            match const_arg.layout.ty.kind {
+                                ty::Float(FloatTy::F32) => {
+                                   if const_arg.to_scalar()?.to_f32()?.is_zero() {
```

Comment:
What about NaN*0? Or infinity?

Reply:
I was hesitating to add the floating point case, I figured there would be some edge case I forgot about... I'll remove this case entirely.